skbio/tree/_bme.py

```diff
@@ -0,0 +1,352 @@
+    ...                      [0.55,  0.55,  0.54,  0.62,  0]],
+    ...                      ['human','monkey','pig','rat','chicken'])
+
+    Perform Balanced Minimum Evoltuion (BME) and construct the minimum evolution tree
+    representing the relationship between those taxa. This is returned as a TreeNode
+    object.
+
+    >>> tree = bme(dm)
+    >>> print(tree.ascii_art())
+                        /-monkey
+    -human--- /--------|
+                       |          /-pig
+                        \--------|
+                                 |          /-chicken
+                                  \--------|
+                                            \-rat
+
+    Notice that, unlike neighbor joining, the tree is rooted at a taxa/leaf node.
+    This will allow it to have nearest neighbor interchange performed on it without
+    needing to re-root the tree.
+
+    """
+    if dm.shape[0] < 3:
+        raise ValueError(
+            "Distance matrix must be at least 3x3 to "
+            "generate a minimum evolution tree."
+        )
+
+    # count the number of taxa
+    n = len(dm.ids)
+
+    # create initial nodes
+    root_name, node_a_name, node_b_name = dm.ids[0], dm.ids[1], dm.ids[2]
+    root = TreeNode(root_name)
+    root_child = TreeNode("")
+    node_a = TreeNode(node_a_name)
+    node_b = TreeNode(node_b_name)
+    root.append(root_child)
+    root_child.append(node_a)
+    root_child.append(node_b)
+
+    # loop over rest of nodes for k = 4 to n
+    for k in range(3, n):
+        taxa = root.count(tips=True) + 1
+        # calculate average distance matrix between subtrees of T_(k-1)
+        # Note that this could be replaced with an updating step to an
+        # initially computed average distance matrix, as noted in GME,
+        # however the updating step requires all subtrees to be updated
+        # in contrast to GME, so instead the matrix is recomputed after
+        # every iteration.
+        adm = _balanced_average_matrix(root, dm)
+        # create node for taxa k
+        node_k_name = dm.ids[k]
+        connecting_node = TreeNode("")
+        node_k = TreeNode(node_k_name)
+        connecting_node.append(node_k)
+        # connecting_node = TreeNode.read(["(%s);" % (node_k_name)])
+        # node_k = connecting_node.find(node_k_name)
+        # create average distance lists for subtrees of T_(k-1)
+        ordered = list(root.postorder(include_self=False))
+        lowerlist = _balanced_lower(ordered, node_k, dm)
+        upperlist = _balanced_upper(ordered, node_k, taxa, dm)
+        # Initialize the list of edge parent nodes for computation.
+        # The tuple is a pair of a TreeNode object and the value of the tree
+        # length for attaching at the edge consisting of the node and its parent.
+        # Here, we start with setting the initial length value of the edge
+        # defined by the root and its unique descendant to 0.
+        edge_list = [(root_child, 0)]
+        for a, i in edge_list:
```

Comment:
This code is risky. You are extending a list while iterating it. Are you sure it works as expected? (Maybe yes, but just double-check).

Reply:
I agree that this code is not best practice and prone to error. My idea is to instead use a separate list of all non-tip/root nodes and then to loop through this list to compute the corresponding tuples for edge_list. This will avoid having to mutate a list during iteration.

Reply:
@christapo Mutating a list while visiting may not always be a bad practice, as long as it is done right. For example [here](https://github.com/scikit-bio/scikit-bio/blob/e6417514c47fce9e140e3017d7c9c7f7e4ca1728/skbio/tree/_tree.py#L697). But in your case, I fear that it will fall into an infinite loop, as demonstrated in the following code:

```python
lst = [1, 2, 3]
for x in lst:
    lst.append(x * 2)
```

If you did not encounter infinite loop, then maybe this piece of code wasn't trigger at all. You may want to double check to see if that's the case.

To make this code safe, you can iterate the list by a pre-defined range of indices, such as:

```python
for k in range(len(edge_list)):
    a, i = edge_list[k]
```

Reply:
@christapo On another thought, please take my previous words with caution. I see in your code that there is `if a.is_tip(): continue`, meaning that the iteration may not infinitely extend `edge_list`. If you are certain that the current code is correct, you don't need to change it.